### PR TITLE
Added services to start and stop the camera

### PR DESCRIPTION
**Pull request by rjingar**
**Please ensure the above _guidelines for contributing_ are met.**

Fixes Issue: #85 

Changes proposed in this pull request:
- Added service "camera/set_power < true/false > " to turn camera on/off.
- Added service "camera/force_power < true/false > " to turn camera on/off regardless of subscribers.
- Added service "camera/is_powered" to check whether camera is powered on or off.
- Added unit tests for the newly added services
- Updated readme

### CODE REVIEW

File: realsense_camera/src/base_nodelet.cpp [hunk 1]
```diff
@@ -284,6 +287,98 @@ namespace realsense_camera
...
+  }
+
+
+  /*
+   * Force Power Camera service
+   */
+  bool BaseNodelet::forcePowerCameraService(realsense_camera::ForcePower::Request & req,
+      realsense_camera::ForcePower::Response & res)
+  {
+    if(req.power_on == true)
```
Comment (mdhorn) [inline realsense_camera/src/base_nodelet.cpp]: Could we just shorten this to:
`start_camera_ = req.power_on;`

File: realsense_camera/src/base_nodelet.cpp [hunk 2]
```diff
@@ -284,6 +287,98 @@ namespace realsense_camera
   }
 
   /*
+   * Check for Nodelet subscribers
+   */
+  bool BaseNodelet::checkForSubscriber()
+  {
+    for(int index=0; index < STREAM_COUNT; index++)
```
Comment (reaganlo) [inline realsense_camera/src/base_nodelet.cpp]: Replace **for(int** with **for (int**

File: realsense_camera/src/base_nodelet.cpp [hunk 3]
```diff
@@ -284,6 +287,98 @@ namespace realsense_camera
...
+    }
+    if (pointcloud_publisher_.getNumSubscribers() > 0)
+    {
+        return true;
+    }
+    return false;
+  }
+
+  /*
+   * Service to check is camera is powered on or not
```
Comment (reaganlo) [inline realsense_camera/src/base_nodelet.cpp]: Update comment  "Service to check **is**" with "Service to check **if**"

File: realsense_camera/src/base_nodelet.cpp [hunk 4]
```diff
@@ -284,6 +287,98 @@ namespace realsense_camera
...
+      return true;
+  }
+
+  /*
+   * Set Power Camera service
+   */
+  bool BaseNodelet::setPowerCameraService(realsense_camera::SetPower::Request & req,
+      realsense_camera::SetPower::Response & res)
+  {
+    if(req.power_on == true)
```
Comment (reaganlo) [inline realsense_camera/src/base_nodelet.cpp]: Replace "if(req.power_on" with "if (req.power_on"

File: realsense_camera/src/base_nodelet.cpp [hunk 5]
```diff
@@ -582,6 +682,17 @@ namespace realsense_camera
   {
     while (ros::ok())
     {
+      if(start_stop_srv_called_ == true && start_camera_ == true)
```
Comment (mdhorn) [inline realsense_camera/src/base_nodelet.cpp]: The logic might read more clearly as:

```
if (start_stop_srv_called_ == true)
{ 
  if (start_camera_ == true)
  { 
    ROS_INFO_STREAM(nodelet_name_ << " - " << startCamera());
    start_stop_srv_called_ = false;
  }
  else 
  {
    ROS_INFO_STREAM(nodelet_name_ << " - " << stopCamera());
    start_stop_srv_called_ = false;
  }
}
```

File: realsense_camera/src/base_nodelet.cpp [hunk 6]
```diff
@@ -284,6 +287,98 @@ namespace realsense_camera
...
+      realsense_camera::SetPower::Response & res)
+  {
+    if(req.power_on == true)
+    {
+      start_camera_ = true;
+      start_stop_srv_called_ = true;
+    }
+    else
+    {
+      if(rs_is_device_streaming(rs_device_, 0) == 0)
```
Comment (reaganlo) [inline realsense_camera/src/base_nodelet.cpp]: Replace "if(rs_is_device_streaming" with "if (rs_is_device_streaming"

File: realsense_camera/src/base_nodelet.cpp [hunk 7]
```diff
@@ -284,6 +287,98 @@ namespace realsense_camera
...
+      {
+        ROS_INFO_STREAM(nodelet_name_ << " - Camera is already Stopped");
+      }
+      else
+      {
+        if(checkForSubscriber() == false)
```
Comment (reaganlo) [inline realsense_camera/src/base_nodelet.cpp]: Replace "if(checkForSubscriber()" with "if (checkForSubscriber()"

File: realsense_camera/src/base_nodelet.cpp (hunk 4 above)
Comment (mdhorn) [inline realsense_camera/src/base_nodelet.cpp]: Please add a space between if and parentheses `if (`

File: realsense_camera/src/base_nodelet.cpp (hunk 6 above)
Comment (mdhorn) [inline realsense_camera/src/base_nodelet.cpp]: Please add a space between if and parentheses `if (`

File: realsense_camera/src/base_nodelet.cpp (hunk 7 above)
Comment (mdhorn) [inline realsense_camera/src/base_nodelet.cpp]: Please add a space between if and parentheses `if (`

**Comment by rjingar** [comment]
@mdhorn @reaganlo updated code based on review comments.

File: realsense_camera/README.md [hunk 8]
```diff
@@ -90,6 +90,12 @@ IR2 camera: Available only for <b>R200</b> cameras.
 ####Services
     get_settings (camera/get_settings)
         Gets the current value of the supported camera options in "options:value" format separated by semicolon.
+    set_power < true/false > (camera/set_power < true/false >) # true - to power on the camera, false - to power off the camera
+        Sets camera power to trun camera on/off. It turns off the camera only if camera has no subscribers. Otherwise it returns false.
+    force_power < true/false > (camera/force_power < true/false >) # true - to power on the camera, false - to forcefully power off the camera
+        Sets camera power to trun camera on/off. It turns off the camera regardless of number of subscribers.
```
Comment (reaganlo) [inline realsense_camera/README.md]: Spellcheck - **trun**

File: realsense_camera/README.md [hunk 9]
```diff
@@ -90,6 +90,12 @@ IR2 camera: Available only for <b>R200</b> cameras.
 ####Services
     get_settings (camera/get_settings)
         Gets the current value of the supported camera options in "options:value" format separated by semicolon.
+    set_power < true/false > (camera/set_power < true/false >) # true - to power on the camera, false - to power off the camera
+        Sets camera power to trun camera on/off. It turns off the camera only if camera has no subscribers. Otherwise it returns false.
```
Comment (reaganlo) [inline realsense_camera/README.md]: Spellcheck - **trun**

File: realsense_camera/src/base_nodelet.cpp [hunk 10]
```diff
@@ -284,6 +287,91 @@ namespace realsense_camera
...
+        }
+        else
+        {
+          ROS_INFO_STREAM(nodelet_name_ << " - Cannot stop the camera. Nodelet has subscriber.");
+          res.success = false;
+          return false;
+        }
+      }
+    }
+    res.success = true;
```
Comment (mdhorn) [inline realsense_camera/src/base_nodelet.cpp]: Move to top of function; assume default will be true.

File: realsense_camera/src/base_nodelet.cpp [hunk 11]
```diff
@@ -284,6 +287,91 @@ namespace realsense_camera
...
+    return true;
```
Comment (mdhorn) [inline realsense_camera/src/base_nodelet.cpp]: change to 
`return res.success`

File: realsense_camera/src/base_nodelet.cpp [hunk 12]
```diff
@@ -284,6 +287,91 @@ namespace realsense_camera
...
+        if (checkForSubscriber() == false)
+        {
+          start_camera_ = false;
+          start_stop_srv_called_ = true;
+        }
+        else
+        {
+          ROS_INFO_STREAM(nodelet_name_ << " - Cannot stop the camera. Nodelet has subscriber.");
+          res.success = false;
+          return false;
```
Comment (mdhorn) [inline realsense_camera/src/base_nodelet.cpp]: Remove the multiple returns; use common return at end of function.

**Comment by mdhorn** [comment]
👍  Approved

**Comment by reaganlo** [comment]
Approved
